File: cpp_merge/classified/packages/plugin-stagehand/stagehand-server/src/session-manager.cpp
#include "session-manager.h"

SessionManager::SessionManager(std::shared_ptr<Logger> logger_, std::shared_ptr<PlaywrightInstaller> playwrightInstaller_) : logger(logger_), playwrightInstaller(playwrightInstaller_)  {
}

std::shared_ptr<Promise<std::shared_ptr<BrowserSession>>> SessionManager::createSession(string sessionId, string clientId)
{
    if (!this->playwrightInstaller->isReady()) {
        try
        {
            std::async([=]() { this->playwrightInstaller->ensurePlaywrightInstalled(); });
        }
        catch (const any& error)
        {
            this->logger->error(std:("Failed to install Playwright:"), error);
            throw any(std::make_shared<Error>(std:("Playwright is not installed and installation failed. Please install Playwright manually.")));
        }
    }
    auto clientSessions = this->getClientSessions(clientId);
    if (clientSessions->get_length() >= this->maxSessionsPerClient) {
        auto oldestSession = const_(clientSessions->sort([=](auto a, auto b) mutable
        {
            return a->createdAt->getTime() - b->createdAt->getTime();
        }
        ))[0];
        if (oldestSession) {
            std::async([=]() { this->destroySession(oldestSession->id); });
        }
    }
    this->logger->info(std:("Creating session ") + sessionId + std:(" for client ") + clientId + string_empty);
    auto env = (process->env->BROWSERBASE_API_KEY) ? std:("BROWSERBASE") : std:("LOCAL");
    auto config = object{
        object::pair{std:("env"), std:("env")}, 
        object::pair{std:("headless"), process->env->BROWSER_HEADLESS != std:("false")}
    };
    if (process->env->BROWSERBASE_API_KEY) {
        config["apiKey"] = process->env->BROWSERBASE_API_KEY;
        config["projectId"] = process->env->BROWSERBASE_PROJECT_ID;
        config["browserbaseSessionCreateParams"] = object{
            object::pair{std:("projectId"), process->env->BROWSERBASE_PROJECT_ID}, 
            object::pair{std:("browserSettings"), object{
                object::pair{std:("blockAds"), true}, 
                object::pair{std:("viewport"), object{
                    object::pair{std:("width"), 1280}, 
                    object::pair{std:("height"), 720}
                }}
            }}
        };
    }
    if (process->env->OLLAMA_BASE_URL) {
        config["modelName"] = OR((process->env->OLLAMA_MODEL), (std:("llama3.2-vision")));
        config["modelBaseUrl"] = process->env->OLLAMA_BASE_URL;
    } else if (process->env->OPENAI_API_KEY) {
        config["modelName"] = std:("gpt-4o");
        config["openaiApiKey"] = process->env->OPENAI_API_KEY;
    } else if (process->env->ANTHROPIC_API_KEY) {
        config["modelName"] = std:("claude-3-5-sonnet-latest");
        config["anthropicApiKey"] = process->env->ANTHROPIC_API_KEY;
    }
    auto stagehand = std::make_shared<Stagehand>(config);
    std::async([=]() { stagehand->init(); });
    auto session = object{
        object::pair{std:("id"), sessionId}, 
        object::pair{std:("clientId"), std:("clientId")}, 
        object::pair{std:("stagehand"), std:("stagehand")}, 
        object::pair{std:("createdAt"), std::make_shared<Date>()}
    };
    this->sessions->set(sessionId, session);
    return session;
}

any SessionManager::getSession(string sessionId)
{
    return this->sessions->get(sessionId);
}

std::shared_ptr<Promise<void>> SessionManager::destroySession(string sessionId)
{
    auto session = this->sessions->get(sessionId);
    if (session) {
        this->logger->info(std:("Destroying session ") + sessionId + string_empty);
        try
        {
            std::async([=]() { session->stagehand->close(); });
        }
        catch (const any& error)
        {
            this->logger->error(std:("Error closing session ") + sessionId + std:(":"), error);
        }
        this->sessions->delete(sessionId);
    }
    return std::shared_ptr<Promise<void>>();
}

array<std::shared_ptr<BrowserSession>> SessionManager::getClientSessions(string clientId)
{
    return Array->from(this->sessions->values())->filter([=](auto session) mutable
    {
        return session->clientId == clientId;
    }
    );
}

std::shared_ptr<Promise<void>> SessionManager::cleanupClientSessions(string clientId)
{
    auto sessions = this->getClientSessions(clientId);
    for (auto& session : sessions)
    {
        std::async([=]() { this->destroySession(session->id); });
    }
    return std::shared_ptr<Promise<void>>();
}

std::shared_ptr<Promise<void>> SessionManager::cleanup()
{
    this->logger->info(std:("Cleaning up all sessions..."));
    for (auto& sessionId : this->sessions->keys())
    {
        std::async([=]() { this->destroySession(sessionId); });
    }
    return std::shared_ptr<Promise<void>>();
}

File: classified/packages/plugin-stagehand/stagehand-server/src/session-manager.ts
import { Stagehand } from '@browserbasehq/stagehand';
import { Logger } from './logger.js';
import { PlaywrightInstaller } from './playwright-installer.js';

export interface BrowserSession {
  id: string;
  clientId: string;
  stagehand: Stagehand;
  createdAt: Date;
}

export class SessionManager {
  private sessions: Map<string, BrowserSession> = new Map();
  private maxSessionsPerClient = 3;

  constructor(
    private logger: Logger,
    private playwrightInstaller: PlaywrightInstaller
  ) {}

  async createSession(sessionId: string, clientId: string): Promise<BrowserSession> {
    // Ensure Playwright is installed before creating session
    if (!this.playwrightInstaller.isReady()) {
      try {
        await this.playwrightInstaller.ensurePlaywrightInstalled();
      } catch (error) {
        this.logger.error('Failed to install Playwright:', error);
        throw new Error(
          'Playwright is not installed and installation failed. Please install Playwright manually.'
        );
      }
    }

    // Check session limit for client
    const clientSessions = this.getClientSessions(clientId);
    if (clientSessions.length >= this.maxSessionsPerClient) {
      // Remove oldest session
      const oldestSession = clientSessions.sort(
        (a, b) => a.createdAt.getTime() - b.createdAt.getTime()
      )[0];
      if (oldestSession) {
        await this.destroySession(oldestSession.id);
      }
    }

    this.logger.info(`Creating session ${sessionId} for client ${clientId}`);

    // Determine environment based on available API keys
    const env = process.env.BROWSERBASE_API_KEY ? 'BROWSERBASE' : 'LOCAL';

    // Create Stagehand configuration
    const config: any = {
      env,
      headless: process.env.BROWSER_HEADLESS !== 'false',
    };

    // Add API keys if available
    if (process.env.BROWSERBASE_API_KEY) {
      config.apiKey = process.env.BROWSERBASE_API_KEY;
      config.projectId = process.env.BROWSERBASE_PROJECT_ID;
      config.browserbaseSessionCreateParams = {
        projectId: process.env.BROWSERBASE_PROJECT_ID!,
        browserSettings: {
          blockAds: true,
          viewport: {
            width: 1280,
            height: 720,
          },
        },
      };
    }

    // Configure LLM provider
    if (process.env.OLLAMA_BASE_URL) {
      // Use Ollama if available
      config.modelName = process.env.OLLAMA_MODEL || 'llama3.2-vision';
      config.modelBaseUrl = process.env.OLLAMA_BASE_URL;
    } else if (process.env.OPENAI_API_KEY) {
      config.modelName = 'gpt-4o';
      config.openaiApiKey = process.env.OPENAI_API_KEY;
    } else if (process.env.ANTHROPIC_API_KEY) {
      config.modelName = 'claude-3-5-sonnet-latest';
      config.anthropicApiKey = process.env.ANTHROPIC_API_KEY;
    }

    const stagehand = new Stagehand(config);
    await stagehand.init();

    const session: BrowserSession = {
      id: sessionId,
      clientId,
      stagehand,
      createdAt: new Date(),
    };

    this.sessions.set(sessionId, session);
    return session;
  }

  getSession(sessionId: string): BrowserSession | undefined {
    return this.sessions.get(sessionId);
  }

  async destroySession(sessionId: string): Promise<void> {
    const session = this.sessions.get(sessionId);
    if (session) {
      this.logger.info(`Destroying session ${sessionId}`);
      try {
        await session.stagehand.close();
      } catch (error) {
        this.logger.error(`Error closing session ${sessionId}:`, error);
      }
      this.sessions.delete(sessionId);
    }
  }

  getClientSessions(clientId: string): BrowserSession[] {
    return Array.from(this.sessions.values()).filter((session) => session.clientId === clientId);
  }

  async cleanupClientSessions(clientId: string): Promise<void> {
    const sessions = this.getClientSessions(clientId);
    for (const session of sessions) {
      await this.destroySession(session.id);
    }
  }

  async cleanup(): Promise<void> {
    this.logger.info('Cleaning up all sessions...');
    for (const sessionId of this.sessions.keys()) {
      await this.destroySession(sessionId);
    }
  }
}
